### ISSUE TITLE
Static Decryption Key?

### ISSUE THREAD

**Issue by Ne0nd0g** [cretz/gopaque#2]
I was recently using this library and noticed that I always get back the same decryption key even when switching up the input parameters when calling `gopaque.CryptoDefault.NewKeyFromReader`. Additionally, I always get back the same "decrypted" EnvU even when using different input keys when calling `gopaque.CryptoDefault.AuthDecrypt`. I will say that crypto isn't my strong suite so I'm not able to tell if I'm just overlooking something or if this is a bug.

I created a sample Go program as a means of testing out my hypothesis. Every time you run the script, the "Original Decryption Key" is the same. I would think that it would be different between runs. 

I created a decryption key using `[]byte("abcdefghijklmnopqrstuvwxyz012345")` and the resulting key was the same as the original key. However, I create a key from scratch without using gopaque and got a different key. The interesting part is the decryption with new key from scratch returns the same results as when using the original key. The program I was using to test this with is here: [main.go.txt](https://github.com/cretz/gopaque/files/3188884/main.go.txt). I've included sample output from the program I was using to test with below:
```
Registered user private key:
:07cd20080168b69273361132924051a020a00fe28f8a343052a96e1e3221a502

Creating new OPRFUserStep1 value with a password of 'neverGonnaGiveYouUp'

Original OPRFUserStep1 value:   fb655527c220c2cc4f3087dedb70923a30a7564574a51cccd8a9afb29b5e9f00
New OPRFUserStep1 value:        2e707040b20e00afb23024f1495dbbaab2b0a8fd9e802f47c21f371c4323b300

Original OPRFUserStep3 value:   [20 128 202 27 56 98 27 228 204 87 147 132 19 111 204 236 219 184 18 221 24 94 114 176 240 85 248 19 91 102 150 216]
New OPRFUserStep3 value:        [154 79 192 230 139 200 33 219 142 13 111 251 73 67 118 208 193 125 227 223 225 190 31 27 42 202 182 115 163 142 171 192]

Creating a RwdU using 'abcdefghijklmnopqrstuvwxyz012345'

Original Decryption Key:        af22e0f057b9dccd4b1be5ce77e2e7d557b57970b5267a90f57960924a87f106
New Decryption Key:             af22e0f057b9dccd4b1be5ce77e2e7d557b57970b5267a90f57960924a87f106
Random decryption key:          af22e0f057b9dccd4b1be5ce77e2e7d557b57970b5267a90f57960924a87f106
Key made from scratch:          4cd6d3f4b44fb16e7dbe34d8d43515707714fdcddf769d8e8486c3090e143607

Decrypted EnvU using original key:      [7 205 32 8 1 104 182 146 115 54 17 50 146 64 81 160 32 160 15 226 143 138 52 48 82 169 110 30 50 33 165 2 161 11 237 12 87 115 84 166 117 143 178 221 88 235 236 32 29 197 63 162 215 57 113 91 247 183 57 78 214 152 92 179]

Decrypted EnvU using key from scrach:   [7 205 32 8 1 104 182 146 115 54 17 50 146 64 81 160 32 160 15 226 143 138 52 48 82 169 110 30 50 33 165 2 161 11 237 12 87 115 84 166 117 143 178 221 88 235 236 32 29 197 63 162 215 57 113 91 247 183 57 78 214 152 92 179]
```

**Comment by cretz** [comment]
Thanks for the report! Sounds a bit scary, heh. I haven't dug, but one of the features is to get the same key each time, but that's probably not what you're saying here. I'm a little busy w/ some other things at the moment, so may not be able to debug for a bit. But I should definitely make a unit test for any error we see here. I'll let you know when I can hop on this and dig into your issue here.

**Comment by Ne0nd0g** [comment]
Thanks for following up, I really appreciate it. I've been continuing to troubleshoot to see if I can figure out if I'm the problem or if there is a bug. 

It seems like no matter what you give the `DeriveKey` function as input, you always get back `af22e0f057b9dccd4b1be5ce77e2e7d557b57970b5267a90f57960924a87f106` and that value is what is ultimately used to decrypt EnvU.

**Comment by Ne0nd0g** [comment]
Created another Go program to run through a couple of tests. I put the program in a gist if you want to reference it. https://gist.github.com/Ne0nd0g/60b8df894a7879cb6d04ea27e87a7e2a

**Comment by Ne0nd0g** [comment]
This test probably most succinctly illustrates the issue and is based on the `crypto_test.go` file:

```
### Encrypt/Decrypt w/ different keys - Test 1 ###
Encryption key:         1fe33b6926e6296d8bff53165f7720d22451ca81c15dc767f739bc5286f2d602
Decryption key:         947c409e640b2d0d1cda2f51dfa45af0f2ea62fb53464f1169c271615eded301
Plain text:             Hack the Planet!
Encrypted bytes:        [51 54 141 136 52 172 77 144 33 173 223 162 123 50 12 152 68 155 132 86 175 81 56 59 21 88 6 72 205 234 124 194 234 125 56 52 228 228 14 183 90 62 151 108 223 154 66 82 87 160 52 17 197 125 8 154 107 215 225 1 68 54 179 46 158 174 112 61 30 224 11 69 13 193 195 169 60 250 55 216]
Decrytped value:        Hack the Planet!

### Encrypt/Decrypt w/ different keys - Test 2 ###
Encryption key:         3df664cb309cccc119cb83eb71da6b9124491a3616bc646ffbc7d1d827e74508
Decryption key:         e5ae15bde40c198033f035437e8dbf42b36933276fcf2243900d64692812490d
Plain text:             Hack the Planet!
Encrypted bytes:        [69 145 228 255 215 51 247 58 99 217 120 95 92 234 167 44 170 223 28 90 248 182 169 21 34 146 29 234 200 97 61 95 67 135 87 44 8 168 90 121 14 228 195 92 221 18 192 202 246 11 69 7 57 212 99 217 57 25 99 124 57 210 18 136 40 107 177 251 128 86 206 193 187 119 21 183 113 167 5 17]
Decrytped value:        Hack the Planet!

### Encrypt/Decrypt w/ different keys - Test 3 ###
Encryption key:         f5f09f467f1a745dcd7b8421f2c2ce945504530557073c66093edc9032b81e07
Decryption key:         6c4095f6ee31776786fd6581f9ae91d3548216321f6e0627a6f56935c28c2101
Plain text:             Hack the Planet!
Encrypted bytes:        [39 60 58 158 165 183 31 236 115 60 161 16 15 200 104 188 123 175 102 145 232 38 162 112 248 90 125 204 114 213 31 136 43 19 131 83 255 219 184 122 66 139 165 132 2 245 207 158 213 215 102 231 60 47 37 180 218 234 132 115 162 38 88 204 31 13 251 198 113 250 160 255 81 87 152 2 213 65 26 91]
Decrytped value:        Hack the Planet!
```

**Comment by cretz** [comment]
Well, that's super scary, heh, my warning/disclaimer is definitely warranted. Can't check in depth yet, but I fear something is wrong with `DeriveKeyHKDF`, unsure. I will try to carve out some time this weekend.

**Comment by ivn-nz** [comment]
I had the same issue, in which regardless the password used during Registration phase and Authentication phase, the program always decrypts fine. I dug a bit on why I was always getting the same key despite being different. I found that function XORKeyStream in util.go sets all values to zero. Then, both the encryptionKey and macKey from func DeriveKey() is always the same.

I copied an XORKeyStream function from kyber library in order to test an alternative. It seems it works fine now.

I guess there is a problem with XORKeyStream, it looks like to me that the function XORs the same bytes, thus the output is 0. I think it should be XORing with a key. 

Note, I am not a cryptographer :).

**Comment by cretz** [comment]
Nice! If you had the time, I'd welcome a PR (and even a test or two if you can). I don't really work on this library any more, and since the code in the lib is downright flawed and dangerous, without an update I might just kill it. If you can't send a PR, would you welcome a link from this repo to your fork?

**Comment by Ne0nd0g** [comment]
What other items are downright flawed and dangerous?

**Comment by cretz** [comment]
None that I know of, just this glaring dangerous flaw.

**Comment by ivn-nz** [comment]
(off-topic): I was looking for an OPAQUE library for a project I am doing, and only found two implementations. I really appreciate the work you've done on that. That library have almost all the protocol implemented from the RFC, although there are some minors things to be finished. tbh, I think that for testing and playing around is great. Even more, i would give it a try even for more serious stuff once this minor flaw is fixed.

If I find the time, i'll do my best to provide a solution for that. I have to consider if your XORKeyStream can be fixed in a way, or if it is better to follow a similar approach of others libraries such as the one in kyber.

**Comment by ivn-nz** [comment]
I am a bit inexperience on go testing and modules. I tested it in my fork, and a simple solution seems to work fine. It is advisable though to have kyber/v3 in place for that.

However, this simple solution in that library versioning is not possible. But,  the XORKeyStream can be modified as I said before, by XORing the key. I want to pass a test, but as the library imports cretz/gopaque from Github it does not take the XORKeyStream from the file locally and therefore fails. I think that if the change on util.go XORKeyStream is pushed, the test should be ok.

The test is that one:
https://gist.github.com/ivn-nz/574854b5753ebd71a7e52e7081d5486c

Let me write here the minor change, new XORKeyStream:
```
func (r *readerStream) XORKeyStream(dst, src []byte) {
    l := len(src)
    if len(dst) < l {
        panic("dst too short")
    }

    buffKey := make([]byte, 32)
    _, err := io.ReadFull(r, buffKey)
    if err != nil {
        panic("reader failed")
    }

    for i := 0; i < l; i++ {
        dst[i] = src[i] ^ buffKey[i]
    }
}
```

**Comment by cretz** [comment]
@ivn-nz - Take a peek at #4 - if you work off that PR/fork and update Kyber, I'll gladly accept the PR...I say your code lgtm

**Comment by cretz** [comment]
I have merged that PR and tagged v0.1.0 ...thanks for the help!
